### PR TITLE
Smart VM

### PR DESCRIPTION
**Pull request by chfast**
New strategy for SmartVM: JIT worker in the background.

### CODE REVIEW

File: eth/main.cpp [hunk 1]
```diff
@@ -1441,8 +1441,10 @@ int main(int argc, char** argv)
 			}
 		}
 #if ETH_EVMJIT
-		else if (arg == "-J" || arg == "--jit")
-			jit = true;
+		else if (arg == "--vm=jit")
```
Comment (gavofyork) [inline eth/main.cpp]: should be `"--vm" "jit"` for consistency with everything else that takes a param e.g. `"--verbosity" "4"`

File: ethvm/main.cpp [hunk 2]
```diff
@@ -43,11 +43,10 @@ void help()
 		<< "    --gas-price <n>  Transaction's gas price' should be <n> (default: 0)." << endl
 		<< "    --sender <a>  Transaction sender should be <a> (default: 0000...0069)." << endl
 		<< "    --origin <a>  Transaction origin should be <a> (default: 0000...0069)." << endl
-#if ETH_EVMJIT || !ETH_TRUE
+#if ETH_EVMJIT || !ETH_TRUETRUE
```
Comment (gavofyork) [inline ethvm/main.cpp]: eh?

File: libevm/SmartVM.cpp [hunk 3]
```diff
@@ -39,30 +45,97 @@ namespace
...
 
 bytesConstRef SmartVM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
 {
-	auto codeHash = sha3(_ext.code);
+	auto codeHash = _ext.codeHash;
 	auto vmKind = VMKind::Interpreter; // default VM
 
 	// Jitted EVM code already in memory?
-	if (evmjit::JIT::isCodeReady(eth2jit(codeHash)))
+	if (evmjit::JIT::isCodeReady(eth2jit(codeHash))) // FIXME: JIT::isCodeReady is not thread-safe
```
Comment (gavofyork) [inline libevm/SmartVM.cpp]: sounds dangerous, especially for AZ which actually does run VM executions in parallel.

File: test/TestHelper.cpp [hunk 4]
```diff
@@ -728,7 +728,7 @@ Options::Options()
 	for (auto i = 0; i < argc; ++i)
 	{
 		auto arg = std::string{argv[i]};
-		if (arg == "--jit")
+		if (arg == "--jit" || arg == "--vm=jit") // TODO: Remove deprecated option "--jit"
```
Comment (gavofyork) [inline test/TestHelper.cpp]: make sure the `--vm=jit` is consistent with style of other options.

File: ethvm/main.cpp (hunk 2 above)
Comment (chfast) [inline ethvm/main.cpp]: No idea :)

File: libevm/SmartVM.cpp (hunk 3 above)
Comment (chfast) [inline libevm/SmartVM.cpp]: That's fixed. Forgot to remove the comment.

**Comment by chfast** [comment]
All issues fixed.

File: libevm/SmartVM.cpp [hunk 5]
```diff
@@ -39,30 +45,97 @@ namespace
...
+		}
+
+	public:
+		JitWorker() noexcept: m_worker([this]{ work(); })
+		{}
+
+		~JitWorker()
+		{
+			{
+				std::lock_guard<std::mutex> lock{x_mutex};
```
Comment (gavofyork) [inline libevm/SmartVM.cpp]: Use `Guard` type for consistency.

File: libevm/SmartVM.cpp [hunk 6]
```diff
@@ -39,30 +45,97 @@ namespace
...
+				m_finished = true;
+			}
+			m_cv.notify_one();
+			m_worker.join();
+		}
+
+		void push(JitTask&& _task)
+		{
+			{
```
Comment (gavofyork) [inline libevm/SmartVM.cpp]: incorrectly placed `{`

File: libevm/SmartVM.cpp [hunk 7]
```diff
@@ -39,30 +45,97 @@ namespace
...
+		        std::lock_guard<std::mutex> lock(x_mutex);
```
Comment (gavofyork) [inline libevm/SmartVM.cpp]: `Guard`

File: evmjit/libevmjit/JIT.cpp [hunk 8]
```diff
@@ -142,9 +147,37 @@ ExecFunc JITImpl::getExecFunc(h256 const& _codeHash) const
...
 
+ExecFunc JITImpl::compile(byte const* _code, uint64_t _codeSize, h256 const& _codeHash)
+{
+	auto name = hash2str(_codeHash);
+	auto module = Cache::getObject(name);
+	if (!module)
+	{
+		// TODO: Listener support must be redesigned. These should be a feature of JITImpl
+		//listener->stateChanged(ExecState::Compilation);
+		assert(_code || !_codeSize); //TODO: Is it good idea to execute empty code?
```
Comment (chriseth) [inline evmjit/libevmjit/JIT.cpp]: I would say that at least empty code should be supported. You can change it to an array of length 1 containing zero.

Comment (chfast) [inline evmjit/libevmjit/JIT.cpp]: It is supported by JIT. But it is only for VM tests as Executive will not call a VM if there is no code.

**Pull request merged**